[WIP] Decorate PodSpec with initcontainers

stevekuznetsov opened the pull request:
[WIP] Decorate PodSpec with initcontainers

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

Introduce a tool to create a Pod from a Job's PodSpec

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

[WIP] Re-add a canary job using PodSpec decoration

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

/area prow
/kind feature
/cc @BenTheElder @kargakis 
/assign @fejta @cjwagner 

We're finally here... lots of stuff TODO still:

 - [ ] why not put the full `PodSpec` into the `ProwJob` instead of decorating at runtime?
 - [ ] why not update the `ProwJob.Status.BuildID` first and read that when launching the pod, instead of the current flow?
 - [ ] let's bikeshed the user-facing config
 - [ ] how do we expose the artifacts dir to the job? Right not it seems a lot of people use `$WORKSPACE/_artifacts` -- can we symlink that or something?
 - [ ] what do we want the dir structure to look like above the clone tree?
 - [ ] ideally we should not expose the entire `$JOB_SPEC` to tools but give them the one or two bits of information they actually need
 - [ ] should the main test container actually get any of the vars in a decorated case? do the utilities correctly encapsulate everything that uses the vars?
 - [ ] where do we put pod logs for the init containers and sidecar?
 - [ ] update the logs endpoint to grab from the test container

stevekuznetsov commented:
Go `size/XXL` or go home
/woof

kargakis commented on prow/kube/prowjob.go:
```diff
@@ -79,22 +79,74 @@ type ProwJob struct {
+	Decorate bool `json:"decorate,omitempty"`
+	// UtilityImages holds pull specs for utility container
+	// images used to decorate a PodSpec.
+	UtilityImages UtilityImages `json:"utility_images,omitempty"`
```
Switch to a pointer, otherwise omitempty does not work.

kargakis commented on prow/kube/prowjob.go:
```diff
@@ -79,22 +79,74 @@ type ProwJob struct {
+	UtilityImages UtilityImages `json:"utility_images,omitempty"`
+	// GCSConfiguration holds options for pushing logs and
+	// artifacts to GCS from a job.
+	GCSConfiguration GCSConfiguration `json:"gcs_configuration,omitempty"`
```
ditto

kargakis commented on prow/kube/prowjob.go:
```diff
@@ -79,22 +79,74 @@ type ProwJob struct {
+	CloneRefs  string `json:"clonerefs,omitempty"`
+	InitUpload string `json:"initupload,omitempty"`
+	Entrypoint string `json:"entrypoint,omitempty"`
+	Sidecar    string `json:"sidecar,omitempty"`
```
godoc these fields too

stevekuznetsov replied:
They're all the same and the Godoc for the struct is pretty succinct ... but could copy-paste it four times :)

kargakis replied:
Godoc here should explain the use of each image.

kargakis replied:
Also potentially where is the canonical source for building these images.

kargakis commented on prow/config/config.go:
```diff
@@ -31,6 +31,8 @@ import (
 	"gopkg.in/robfig/cron.v2"
 	"k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/test-infra/prow/pod-utils/decorate"
```
nit: wrong group

kargakis commented on prow/cmd/mkpod/main.go:
```diff
@@ -0,0 +1,92 @@
+
+	"github.com/ghodss/yaml"
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/pod-utils/decorate"
```
nit: wrong group

kargakis commented on prow/kube/prowjob.go:
```diff
@@ -79,22 +79,74 @@ type ProwJob struct {
+	// TimeoutMinutes is the number of minutes that
+	// the pod utilities will wait before aborting
+	// a job. Only applicable if decorating the PodSpec.
+	TimeoutMinutes int `json:"timeout_minutes,omitempty"`
```
Did we agree on switching this to a duration or no?

stevekuznetsov replied:
Want some direction on internal/external config for time.Duration since you can't deserialize into it

kargakis commented on prow/pod-utils/decorate/podspec.go:
```diff
@@ -23,36 +23,204 @@ import (
+	LogMountName            = "logs"
+	LogMountPath            = "/logs"
+	CodeMountName           = "code"
+	CodeMountPath           = "/go"
```
Is this ok for non-golang projects?

stevekuznetsov replied:
Figuring out what this should be is one of the TODOs in the body, /cc @BenTheElder

kargakis replied:
/data ? /src ? /code ?

krzyzacy replied:
this should be configurable?

kargakis commented on prow/kube/prowjob.go:
```diff
@@ -79,22 +79,74 @@ type ProwJob struct {
+	// Decorate determines if the user-provided PodSpec
+	// will be decorated with more containers to add in
+	// cloning and logging functionality.
+	Decorate bool `json:"decorate,omitempty"`
```
Is there any value in allowing selective enablement of individual utilities, eg. enabling cloning only?

stevekuznetsov replied:
No, e.g. for cloning you want `initupload` as well else you can't tell if it failed ... I guess you could get those but not the sidecar/entrypoint but I'm not aware of a use-case today for it

kargakis commented:
> why not update the ProwJob.Status.BuildID first and read that when launching the pod, instead of the current flow?

When we launch the pod we already know the build id, right? Anyway sounds like a separate issue if we agree it's an issue?

stevekuznetsov commented:
> Are you asking about generation of the extra config and adding it in config.yaml?

Why not do the transformation at `trigger` time and then `plank` just manages the lifecycle, since there is only one correct way to transform the qusi-spec to a real Pod anyway

> When we launch the pod we already know the build id, right? Anyway sounds like a separate issue if we agree it's an issue?

Sure, I got lost a little bit trying to thread through when each controller updated the status of the PJ

kargakis commented:
> Why not do the transformation at trigger time and then plank just manages the lifecycle, since there is only one correct way to transform the qusi-spec to a real Pod anyway

The only issue I can think of is that the size of prowjobs almost doubles(?) by putting three/four extra containers and there are a lot of components that fetch all of the prowjobs in the namespace including deck which also loads them in the client. We could workaround loading in the client by dropping the extra data with a url parameter (similar fix to https://github.com/kubernetes/test-infra/pull/7386) but the rest of the components that use the API are still affected. I am still waiting on https://github.com/kubernetes/kubernetes/pull/57682 to land in our CI cluster so I can start assessing what's the real cost of prowjob LISTs today. Based on the times our core controllers werent using informers, I expect it to be high.

kargakis commented:
> We could workaround loading in the client by dropping the extra data with a url parameter (similar fix to #7386)

Sorry, this is already fixed. I need to log out.

stevekuznetsov commented:
@cjwagner re: your thumbs down on my comment -- if you see a better way for me to break out commits into logical bits that can be reviewed alone, please let me know. I am doing my best to keep this manageable but the core changes must land together, there is no point in adding only one `initContainer` or something.

stevekuznetsov commented:
@krzyzacy @kargakis @BenTheElder @fejta @cjwagner rebased on top of other PRs, broke up work into many small commits. Hopefully easier to review.

stevekuznetsov commented:
@cjwagner `s/XXL/XL/`

stevekuznetsov commented:
Depending on how many of the TODOs in the OP we care about, this is ready. Some of those are niceties that we can improve on in follow-ups.

kargakis commented on prow/config.yaml:
```diff
@@ -5761,6 +5761,47 @@ presubmits:
           requests:
             memory: "2Gi"
 
+  - name: pull-test-infra-verify-config-canary
```
Seems like this can also be added in a separate PR.

stevekuznetsov replied:
If we care, sure, I just had it here to test the generation

stevekuznetsov replied:
Also shows us literally what the config looks like

kargakis commented on prow/config/config.go:
```diff
@@ -483,6 +503,16 @@ func parseConfig(c *Config) error {
 
 	return nil
 }
```
nit: space plz

kargakis commented on prow/kube/prowjob.go:
```diff
@@ -124,6 +138,19 @@ type ProwJobSpec struct {
+// UtilityImages holds pull specs for the utility images
+// to be used for a job
+type UtilityImages struct {
+	// CloneRefs is the pull spec used for the clonerefs utility
```
I was interested in seeing godoc specific to each field. For example:
```go
// CloneRefs is the image used for cloning all the necessary source code
// inside the test container. Read more about clonerefs in <link to clonerefs readme>
```

kargakis replied:
And the readme should contain info on how to use the utility manually and a reference to how prow is using it. Addings READMEs is not something that should be handled here tho.

cjwagner commented on prow/pod-utils/decorate/podspec.go:
```diff
@@ -23,36 +23,204 @@ import (
+			Log:          cloneLog,
+			GitUserName:  clonerefs.DefaultGitUserName,
+			GitUserEmail: clonerefs.DefaultGitUserEmail,
+			GitRefs:      append(pj.Spec.ExtraRefs, pj.Spec.Refs),
```
This could modify `pj.Spec.ExtraRefs`.

cjwagner commented on prow/pod-utils/decorate/podspec.go:
```diff
@@ -23,36 +23,204 @@ import (
+			TimeoutMinutes: pj.Spec.TimeoutMinutes,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("could not encode sidecar configuration as JSON: %v", err)
```
s/sidecar/entrypoint/

cjwagner commented on prow/config/config.go:
```diff
@@ -507,6 +537,51 @@ func validatePresets(name string, labels map[string]string, spec *v1.PodSpec, pr
+	}
+
+	if len(spec.InitContainers) != 0 {
+		return fmt.Errorf("job %s specified init containers, which is not allowed", name)
```
Should jobs be allowed to specify init containers if they are not using decoration?

stevekuznetsov replied:
Not today -- we need to have a strict paradigm for this sort of thing. It's not supported today, unless we come up with a strict design that allows for it to happen, seems OK to continue to not allow it.

CatherineF-dev replied:
Hi Steve, does "a strict paradigm for this sort of thing" mean the order of initContainers?

stevekuznetsov commented on prow/config.yaml:
```diff
@@ -5761,6 +5761,47 @@ presubmits:
+    path_alias: "k8s.io/test-infra"
+    decorate: true
+    utility_images:
+      clonerefs: "registry.svc.ci.openshift.org/ci/clonerefs:latest"
```
@BenTheElder can we get builds for these images set up on `gcr.io`?

krzyzacy replied:
do you want to add some image push rules for them?

stevekuznetsov replied:
Looks like the rules exist, just have not been used. We can continue to serve from our registry until these are more stable.

fejta commented:
Can we bikeshed design in a google doc? Seems to be an easier way to do that. Kind of painful to do this inside a code review.

fejta commented on prow/config/config.go:
```diff
@@ -276,6 +278,12 @@ func parseConfig(c *Config) error {
 		if v.MaxConcurrency < 0 {
 			return fmt.Errorf("job %s jas invalid max_concurrency (%d), it needs to be a non-negative number", v.Name, v.MaxConcurrency)
 		}
+		if err := validatePodSpec(v.Name, kube.PresubmitJob, v.Spec); err != nil {
```
Can we ensure this has unit test coverage?

In other words I would expect to see something like:
```
cases := []struct {
  name string
  config string // or maybe Config
  err bool
}{
  {
    name: "a config with a job that uses an init container",
    config: `name: foo; spec: initContainers: something`
    err: true,
  },
  // cases for the cross product of job type and validateFoo function failure
}
  for _, tc := range cases { c := yaml.Unmshal(tc.config); actual := parseConfig(&c); actual != tc.expected }
```

Does this already exist somewhere and I'm just missing it?

fejta commented on prow/pod-utils/decorate/podspec_test.go:
```diff
@@ -0,0 +1,329 @@
+			podName: "pod",
+			buildID: "blabla",
+			labels:  map[string]string{"needstobe": "inherited"},
+			pjSpec: kube.ProwJobSpec{
```
I feel like this would be easier for humans to read as input prow yaml, output kubernetes resource yaml, rather than creating structs?

stevekuznetsov replied:
(I personally feel strongly in the other direction)

stevekuznetsov commented:
> Can we bikeshed design in a google doc? Seems to be an easier way to do that. Kind of painful to do this inside a code review.

Previous Google Doc for bike shedding didn't get too much done, but sure, let's try in [this one](https://docs.google.com/document/d/1MHr2c_ckMWyApZzaVQXDhZ5fBmr960eMt_lq6uvEuEw/edit).

@cjwagner @kargakis @krzyzacy @BenTheElder @fejta

stevekuznetsov commented:
Forgot to push that commit earlier -- addressed all comments on the Google Doc. Do we have any other comments? As this is entirely opt-in and does not break existing flows, I say we merge it and iterate later on parts we don't like in retrospect unless people have strong opinions today...

kargakis commented:
Can you validate that UtilityConfig is used only with "agent: Kubernetes" jobs?

kargakis commented on prow/kube/prowjob.go:
```diff
@@ -115,6 +115,20 @@ type ProwJobSpec struct {
 	// of this job that can run in parallel at once
 	MaxConcurrency int `json:"max_concurrency,omitempty"`
 
+	// Decorate determines if the user-provided PodSpec
```
Group TimeoutMinutes with these fields.

kargakis replied:
Thoughts on keeping these fields as is vs grouping into a struct similar to UtilityConfig? Top-level spec fields should be as reusable across agents as possible.

stevekuznetsov replied:
That's the sort of bike-shed I want to have. I think the current design of one spec for many controllers is bad and I'm not sure we want to stick with it forever but if we want to move these, sure, I don't feel too strongly

kargakis replied:
I don't think it's bad, we just don't do unions correctly.

kargakis replied:
Btw there is a proposal to add first-class support for union in k8s

https://github.com/kubernetes/community/pull/388

stevekuznetsov commented:
@kargakis addressed comments, PTAL

kargakis commented on prow/pod-utils/decorate/podspec.go:
```diff
@@ -23,36 +23,211 @@ import (
+	ArtifactsEnv            = "ARTIFACTS"
+	ArtifactsPath           = LogMountPath + "/artifacts"
+	CodeMountName           = "code"
+	CodeMountPath           = "/home/prow/go"
```
Will this be hard to change later? I don't like that we use a go-specific path.

stevekuznetsov replied:
Again this is literally an ephemeral path that can be mounted anywhere and has no meaning outside of the container. It does not exist until the container is started and the mount is created. It should not impact anything anywhere. Nobody should be aware of this path or use it for anything.

stevekuznetsov replied:
And allowing someone to configure it to point somewhere misleads them into thinking that we will somehow use whatever may exist at that path already instead of clobbering it with the volume mount.

kargakis replied:
I wouldnt want to make this configurable. Do you need it to be go-specific?

kargakis commented on prow/kube/prowjob.go:
```diff
@@ -124,11 +116,51 @@ type ProwJobSpec struct {
+	// Decorate determines if the user-provided PodSpec
+	// will be decorated with more containers to add in
+	// cloning and logging functionality.
+	Decorate bool `json:"decorate,omitempty"`
```
Is this needed anymore now with the new struct?

stevekuznetsov replied:
True, could check for non-nil

stevekuznetsov commented:
@kargakis removed the boolean

kargakis commented on prow/pod-utils/decorate/podspec.go:
```diff
@@ -23,36 +23,211 @@ import (
+			Log:          cloneLog,
+			GitUserName:  clonerefs.DefaultGitUserName,
+			GitUserEmail: clonerefs.DefaultGitUserEmail,
+			GitRefs:      refs,
```
nit: 
```go
GitRefs: append(pj.Spec.ExtraRefs, pj.Spec.Refs)
```

stevekuznetsov replied:
Actually that would mutate the storage under `pj.Spec.ExtraRefs` which @cjwagner pointed out which is why I do it this way now :)

kargakis replied:
`append` is not mutating its input slice: https://play.golang.org/p/Wkxv2cKTV3f

stevekuznetsov replied:
https://stackoverflow.com/questions/35920534/why-does-append-modify-passed-slice

kargakis replied:
right

>  If the capacity of s is not large enough to fit the additional values, append allocates a new, sufficiently large underlying array that fits both the existing slice elements and the additional values. Otherwise, append re-uses the underlying array.

kargakis commented:
/approve

Please do some squashing then I can lgtm

stevekuznetsov commented:
@kargakis squashed

stevekuznetsov commented:
Opened follow-up issues to track future work once this is in.

kargakis commented:
/hold cancel
